honeybadgermpc/hbavss_batch.py

```diff
@@ -0,0 +1,235 @@
+from honeybadgermpc.exceptions import HoneyBadgerMPCError
+from honeybadgermpc.protocols.reliablebroadcast import reliablebroadcast
+
+# TODO: Move these to a separate file instead of using it from batch_reconstruction.py
+from honeybadgermpc.batch_reconstruction import subscribe_recv, wrap_send
+
+
+class HbAVSSMessageType:
+    OK = "OK"
+    IMPLICATE = "IMPLICATE"
+    READY = "READY"
+
+
+class HbAvssBatch(object):
+    def __init__(self, public_keys, private_key, g, h, n, t, my_id, send, recv):
+        self.public_keys, self.private_key = public_keys, private_key
+        self.n, self.t, self.my_id = n, t, my_id
+        self.g = g
+        self.poly_commit = PolyCommit(g, h)
+
+        # Create a mechanism to split the `recv` channels based on `tag`
+        self.subscribe_recv_task, self.subscribe_recv = subscribe_recv(recv)
+
+        # Create a mechanism to split the `send` channels based on `tag`
+        def _send(tag):
+            return wrap_send(tag, send)
+        self.get_send = _send
+
+        # This is added to consume the share the moment it is generated.
+        # This is especially helpful when running multiple AVSSes in parallel.
+        self.output_queue = asyncio.Queue()
+
+        self.field = ZR
+        self.poly = polynomials_over(self.field)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.subscribe_recv_task.cancel()
+
+    async def _process_avss_msg(self, avss_id, dealer_id, dispersal_msg):
+        tag = f"{dealer_id}-{avss_id}-AVSS"
+        send, recv = self.get_send(tag), self.subscribe_recv(tag)
+
+        def multicast(msg):
+            for i in range(self.n):
+                send(i, msg)
+
+        commitments, ephemeral_public_key, encrypted_witnesses = loads(
+            dispersal_msg)
+
+        # # all_encrypted_witnesses: n
+        shared_key = pow(ephemeral_public_key, self.private_key)
+
+        shares = [None] * (self.t+1)
+        witnesses = [None] * (self.t+1)
+        # Decrypt
+        for k in range(self.t+1):
+            # print(self.my_id, k, len(encrypted_witnesses[self.my_id * (self.t+1) + k]))
+            shares[k], witnesses[k] = SymmetricCrypto.decrypt(
+                str(shared_key).encode("utf-8"), encrypted_witnesses[self.my_id * (self.t+1) + k])
+
+        ##################### Don't know why this doesn't work #####################
+        # # Decrypt
+        # for k in range(self.t+1):
+        #     shares[k], witnesses[k] = SymmetricCrypto.decrypt(
+        #         str(shared_key).encode("utf-8"), encrypted_witnesses[self.my_id][k])
+        ############################################################################
+
+        # verify & send all ok
+        for i in range(self.t+1):
+            if self.poly_commit.verify_eval(commitments[i], self.my_id+1, shares[i], witnesses[i]):
+                multicast(HbAVSSMessageType.OK)
+            else:
```

Comment:
Since nobody is processing implicate message right now -> it is better to raise an `error` instead of sending the message. This will make debugging easier if something is actually failing.

Reply:
Resolved.